Allow pushing user-allocation membership to Keycloak

QuanMPhm opened the pull request:
Closes https://github.com/nerc-project/operations/issues/948. More details in the commit message
There are still some questions I have below, so this is still a draft for now.

QuanMPhm commented on src/coldfront_plugin_cloud/kc_client.py:
```diff
@@ -0,0 +1,81 @@
+
+    def get_user_id(self, cf_username) -> str | None:
+        """Return None if user not found"""
+        # TODO (Quan): Confirm that Coldfront usernames map to Keycloak emails, not email, or something else?
```
@knikolla @naved001

QuanMPhm commented:
@knikolla Two more questions:
1. Do we also want `validate_allocations` to add PIs to pre-existing allocations? 
2. When a PI adds a user to an Coldfront project or allocation, do those users also get added to a the project's Keycloak group?

knikolla commented on src/coldfront_plugin_cloud/kc_client.py:
```diff
@@ -0,0 +1,81 @@
+    @functools.cached_property
+    def api_client(self):
+        params = {
+            "grant_type": "password",
```
See if you can use the client credentials flow instead of admin password here.

knikolla commented on src/coldfront_plugin_cloud/tasks.py:
```diff
@@ -154,6 +155,17 @@ def set_quota_attributes():
 
         allocator.set_quota(project_id)
 
+        # After setting everything on cluster, add user to Keycloak group
```
This function is growing too long (doing too many things) and I think there might a better place for the code in this block.

Perhaps the base class in `base.py` can abstract some of the duties, among `create_project`, `get_or_create_federated_user` and `assign_role_on_user`.

knikolla commented on src/coldfront_plugin_cloud/openshift.py:
```diff
@@ -364,6 +364,8 @@ def assign_role_on_user(self, username, project_id):
             # Role already exists, ignore
             pass
 
+        super().assign_role_on_user(username, project_id)
```
Instead of the resource allocators, I think this should be handled in https://github.com/nerc-project/coldfront-plugin-cloud/blob/0bfb50570f6e6c25f685e198cf7a52216b706f57/src/coldfront_plugin_cloud/tasks.py#L89

knikolla commented on src/coldfront_plugin_cloud/base.py:
```diff
@@ -45,6 +48,29 @@ def get_or_create_federated_user(self, username):
         return user
 
+    def assign_role_on_user(self, username, project_id):
+        self.kc_admin_client.create_group(project_id)
```
Instead of using the group ID, how about introducing a new Resource Attribute to Resources that accepts a format string. For example defaulting to `"{resource_name}/{project_name}"`

This would allow operator to specify their own format for group names for the clusters.

So you'd read the resource attribute and then call format on the string providing a few documented options for the available variables.

```python
>>> template = "{resource_name}/{project_name}"
>>> group = template.format(**{"resource_name": "NERC-OCP", "project_name": "1234"}
... )
>>> 
>>> group
'NERC-OCP/1234'
```

QuanMPhm replied:
Apologies. Missed this

knikolla reviewed (changes requested):
Did a quick first pass and provided some comments.

Also this needs to be possible configurable via a setting.

knikolla commented:
@QuanMPhm Actually, another thought, do you think it would make sense to implement this in the Keycloak plugin? https://github.com/nerc-project/coldfront-plugin-keycloak

It could listen to signals in the same way that the cloud plugin listens to signals. It already has a keycloak client implemented. 

And there is nothing in pushing users to a Keycloak group that is specific to either OpenShift or OpenStack.

QuanMPhm commented:
@knikolla I see that it does make sense to seperate the Keycloak functionality from the rest of the plugin. It makes sense to me. I forgot that repo existed. There would need to be some overhaul to add integration and unit tests to `coldfront-plugin-keycloak`. Is that fine?

knikolla commented:
> @knikolla I see that it does make sense to seperate the Keycloak functionality from the rest of the plugin. It makes sense to me. I forgot that repo existed. There would need to be some overhaul to add integration and unit tests to `coldfront-plugin-keycloak`. Is that fine?

For now let's keep it here (as not to frontload the work) and we can easily split it out later if needed. Perhaps try implementing it here via signals so as to keep it loosely coupled so that if we need to split it later it doesn't require a lot of uncoupling.

QuanMPhm commented:
@knikolla @larsks I have a question about the Coldfront Keycloak PR. I know we previously decided to represent user-membership to an allocation by adding them to a Keycloak group with the same name as the `ALLOCATION_PROJECT_ID` . Would it make more sense for the group name to be the allocation's ID on Coldfront instead? This has a stronger guarantee of uniqueness, and we won't have to disambiguate for projects with the same name in different clusters. Developers or tools can correlate the allocation id to the on-cluster project id by using the Coldfront API.

knikolla commented:
> @knikolla @larsks I have a question about the Coldfront Keycloak PR. I know we previously decided to represent user-membership to an allocation by adding them to a Keycloak group with the same name as the `ALLOCATION_PROJECT_ID` . Would it make more sense for the group name to be the allocation's ID on Coldfront instead? This has a stronger guarantee of uniqueness, and we won't have to disambiguate for projects with the same name in different clusters. Developers or tools can correlate the allocation id to the on-cluster project id by using the Coldfront API.

See my comment here https://github.com/nerc-project/coldfront-plugin-cloud/pull/249#discussion_r2977091803

QuanMPhm commented:
@knikolla @larsks I have revised the PR and made the Keycloak integration a toggle-able feature. Calls to the Keycloak API is now triggered through signals. I have questions below that I think would be of interest to you both

QuanMPhm commented on src/coldfront_plugin_cloud/tasks.py:
```diff
@@ -128,3 +136,71 @@ def remove_user_from_allocation(allocation_user_pk):
+    project_name = allocation.project.title
+    resource_name = allocation.resources.first().name
+
+    # TODO (Quan) Do we also want some validation steps for the format string at creation time to make sure its valid?
```
@knikolla @larsks And are we fine with the security of `string.Template`?

knikolla replied:
The project name is not the right choice as a Project can have several allocations. This should be the `Allocated Project Name` or `Allocated Project ID`. You could potentially pass to the templater all the attributes associated with an allocation.

QuanMPhm commented on src/coldfront_plugin_cloud/tasks.py:
```diff
@@ -128,3 +136,71 @@ def remove_user_from_allocation(allocation_user_pk):
+    resource_name = allocation.resources.first().name
+
+    # TODO (Quan) Do we also want some validation steps for the format string at creation time to make sure its valid?
+    # Do we also want to include allocation id or something at allocation level?
```
@knikolla @larsks Right now, I just included resource name and Coldfront project name as allowed variables in the template

knikolla commented on src/coldfront_plugin_cloud/signals.py:
```diff
@@ -51,8 +57,14 @@ def activate_allocation_user_receiver(sender, **kwargs):
         add_user_to_allocation(allocation_user_pk)
 
+    if is_keycloak_enabled():
+        add_user_to_keycloak(allocation_user_pk)
```
Why are you not executing these asynchronously?

QuanMPhm replied:
I forgot to ask :P

knikolla replied:
It was safe to assume considering the structure of the function and the overall symmetry.

knikolla commented on src/coldfront_plugin_cloud/kc_client.py:
```diff
@@ -0,0 +1,85 @@
+        if response.status_code not in (201, 409):
+            response.raise_for_status()
+
+    def create_user(self, cf_username):
```
I would move these helper functions in the test rather than place them here.

knikolla reviewed (changes requested):
Quick first pass for direction feedback.

knikolla commented on src/coldfront_plugin_cloud/attributes.py:
```diff
@@ -45,6 +46,7 @@ class CloudAllocationAttribute:
     CloudResourceAttribute(name=RESOURCE_PROJECT_DOMAIN),
     CloudResourceAttribute(name=RESOURCE_ROLE),
     CloudResourceAttribute(name=RESOURCE_QUOTA_RESOURCES),
+    CloudResourceAttribute(name=RESOURCE_KEYSTONE_GROUP_TEMPLATE),
```
Keycloak :)

knikolla reviewed (changes requested):
Looking much better. Almost there.

A follow-up to this (after we merge the `validate_allocations` refactor) should investigate how to cleanly implement this in `validate_allocations` too.

knikolla commented:
> Looking much better. Almost there.
> 
> A follow-up to this (after we merge the `validate_allocations` refactor) should investigate how to cleanly implement this in `validate_allocations` too.

Actually... I would make it a whole separate CLI command. `validate_keycloak_groups`.

QuanMPhm commented:
@knikolla I've allowed the group name template string to accept any allocation attribute. Further documentation is in the docstring for the function `_get_keycloak_group_name`

knikolla commented:
@naved001 would appreciate a pass from you.

knikolla commented on src/coldfront_plugin_cloud/tasks.py:
```diff
@@ -128,3 +140,79 @@ def remove_user_from_allocation(allocation_user_pk):
+        attributes.RESOURCE_KEYCLOAK_GROUP_TEMPLATE
+    )
+    if group_name_template is None:
+        logger.warning(
```
I had some more time to think about this. Instead of logging a warning, log an `info` message here, this way the presence of the Keycloak Group Template Resource Attribute can act as a flag for enabling the feature for each specific Resource individually. 

Therefore also switch the order so that the group template is parsed before you look up the user.

> Keycloak enabled but no group name template specified for resource <>. Skipping addition to Keycloak group.

knikolla commented on src/coldfront_plugin_cloud/tasks.py:
```diff
@@ -128,3 +140,79 @@ def remove_user_from_allocation(allocation_user_pk):
+        attributes.RESOURCE_KEYCLOAK_GROUP_TEMPLATE
+    )
+    if group_name_template is None:
+        logger.warning(
```
Same comment as from addition.

knikolla reviewed (changes requested):
Had some more time to think about how to expose enabling/disable the feature.

naved001 commented on src/coldfront_plugin_cloud/kc_client.py:
```diff
@@ -0,0 +1,73 @@
+        self.base_url = os.getenv("KEYCLOAK_BASE_URL")
+        self.realm = os.getenv("KEYCLOAK_REALM")
+        self.client_id = os.getenv("KEYCLOAK_CLIENT_ID", "coldfront")
+        self.client_secret = os.getenv("KEYCLOAK_CLIENT_SECRET", "nomoresecret")
```
I'd rather we don't set the default id and secret.

naved001 commented on src/coldfront_plugin_cloud/kc_client.py:
```diff
@@ -0,0 +1,73 @@
+            "client_id": self.client_id,
+            "client_secret": self.client_secret,
+        }
+        r = requests.post(self.token_url, data=params).json()
```
maybe some error handling would be useful before you try to access `r["access_token"]`.

QuanMPhm replied:
used `raise_for_status()`

naved001 commented on src/coldfront_plugin_cloud/kc_client.py:
```diff
@@ -0,0 +1,73 @@
+        url = f"{self.base_url}/admin/realms/{self.realm}/users/{user_id}/groups"
+        r = self.api_client.get(url)
+        r.raise_for_status()
+        return [group["name"] for group in r.json()]
```
this may be my limited knowledge of keycloak, but are we guaranteed to have a user be part of some group?

QuanMPhm replied:
I can't say what the production environment will look like, but local testing doesn't suggest users are automatically added to some group. Regardless, I don't think that should be a problem for consumers of the Keycloak API? @knikolla ?

knikolla replied:
It is likely that the user may be part of other groups besides the ones corresponding to respective allocations. This is at least guaranteed for PIs who are part of a group called `pi`.

naved001 commented on src/coldfront_plugin_cloud/tasks.py:
```diff
@@ -128,3 +140,79 @@ def remove_user_from_allocation(allocation_user_pk):
+    Acceptable variables for the group name template string is:
+    - $resource_name: the name of the resource (e.g. "OpenShift")
+    - Any allocation attribute defined for the allocation, with spaces replaced by underscores and
+    all lowercase (e.g. for `Project Name`, the variable would be `$project_name`)
```
why is there a `$` before `project_name`?

QuanMPhm replied:
That is the syntax for [Python string templates](https://docs.python.org/3/library/string.html#template-strings-strings), `$` marks the start of placeholders that will be substituted

naved001 commented on src/coldfront_plugin_cloud/kc_client.py:
```diff
@@ -0,0 +1,73 @@
+    def create_group(self, group_name):
+        url = f"{self.base_url}/admin/realms/{self.realm}/groups"
+        payload = {"name": group_name}
+        response = self.api_client.post(url, json=payload)
```
does this return a group id when a group is created? If yes, we should return that to the caller and we won't need an additional call to get_group_id

QuanMPhm replied:
From local testing and [the online documentation](https://www.keycloak.org/docs-api/latest/rest-api/index.html#_post_adminrealmsrealmgroups), it just returns the status code

naved001 commented on src/coldfront_plugin_cloud/kc_client.py:
```diff
@@ -0,0 +1,73 @@
+        }
+        session = requests.session()
+        session.headers.update(headers)
+        return session
```
We are creating and caching this session with a token (and cache the instantiation of the class again later). When does the keycloak token expire? Just wondering how long we can continue to use this. Or is that not a concern since we'll run this thing as needed and then when its needed we start from scratch . I admit I don't know how coldfront tasks call these.

QuanMPhm replied:
@naved001 The keycloak API will be called whenever Coldfront's [`remove/activate user`](https://github.com/nerc-project/coldfront-plugin-cloud/blob/0bfb50570f6e6c25f685e198cf7a52216b706f57/src/coldfront_plugin_cloud/signals.py#L46-L58) signals are triggered, which happens when a user is add/removed from an allocation. That being said, I would assume the api_client will be called somewhat often, especially during batch user uploads. 

@naved001 @knikolla Do we still want to cache the api client? Or make a follow-up PR to set a TTL mechanism?

knikolla replied:
Just remove the caching for now.

QuanMPhm committed:
Allow pushing user-allocation membership to Keycloak

A Keycloak admin client has been added
When `activate_allocation` is called, the user is added
to a Keycloak group named using a format string defined in the
allocation's resource attribute "Format String for Keystone Group Names"
If the user does not already exist in Keycloak, the case is ignored for now

Keycloak integration is optional, toggled by setting the env var "KEYCLOAK_BASE_URL"
Authentication to Keycloak is done via client credentials grant

When `deactivate_allocation` is called, the user is removed from the Keycloak group

New functional test added for Keycloak integration

A comment in `validate_allocations` has been updated to
reflect the more restrictive validation behavior, where users on cluster projects
will be removed if they are not part of the Coldfront allocation (rather
than if they are not registered on Coldfront at all).

QuanMPhm commented on src/coldfront_plugin_cloud/tasks.py:
```diff
@@ -128,3 +140,79 @@ def remove_user_from_allocation(allocation_user_pk):
+    return template_string.replace(" ", "_").lower()
+
+
+def _get_keycloak_group_name(allocation: Allocation, template_string: str) -> str:
+    """
+    Acceptable variables for the group name template string is:
+    - $resource_name: the name of the resource (e.g. "OpenShift")
+    - Any allocation attribute defined for the allocation, with spaces replaced by underscores and
+    all lowercase (e.g. for `Project Name`, the variable would be `$project_name`)
+    """
+    resource_name = allocation.resources.first().name
+    allocation_attrs_list: list[AllocationAttribute] = (
+        allocation.allocationattribute_set.all()
+    )
+
+    template_sub_dict = {"resource_name": resource_name}
+    for attr in allocation_attrs_list:
+        template_sub_dict[
+            _clean_template_string(attr.allocation_attribute_type.name)
+        ] = attr.value
+
+    return Template(template_string).substitute(**template_sub_dict)
```
@knikolla @naved001 Not raising the error could lead to users being silently added to incorrectly-named Keycloak groups. Is that acceptable?

knikolla replied:
The task should break rather than do the wrong thing.

QuanMPhm commented on src/coldfront_plugin_cloud/signals.py:
```diff
@@ -25,6 +27,10 @@ def is_async():
 
 
+def is_keycloak_enabled():
+    return os.getenv("KEYCLOAK_BASE_URL")
```
I think one env var is enough to indicate Keycloak is enabled. It's the operator's problem if the config is misconfigured

larsks replied:
If all of those variables are required we should be validating that they are set and emitting a useful error message if not. On the other hand, this might be the wrong place for it.

Maybe in `KeyCloakAPIClient`, where you read in the variables to set up the client:

```python
class KeyCloakAPIClient:
    def __init__(self):
        self.base_url = os.getenv("KEYCLOAK_BASE_URL")
        self.realm = os.getenv("KEYCLOAK_REALM")
        self.client_id = os.getenv("KEYCLOAK_CLIENT_ID")
        self.client_secret = os.getenv("KEYCLOAK_CLIENT_SECRET")

```

...but have you considered [pydantic-settings](https://github.com/pydantic/pydantic-settings)?

QuanMPhm replied:
@knikolla Are you fine adding a pydantic settings module in this PR, or hold this off to a PR that puts ALL the settings in the repo into a settings model?

knikolla replied:
I'm fine with pushing this into a separate refactoring PR.

QuanMPhm commented on src/coldfront_plugin_cloud/tasks.py:
```diff
@@ -128,3 +140,79 @@ def remove_user_from_allocation(allocation_user_pk):
+        return
+
+    group_name = _get_keycloak_group_name(allocation, group_name_template)
+    group_id = kc_admin_client.get_group_id(group_name)
```
@knikolla @naved001 This scenario, where the group_id turns out to be `None` will probably happen quite a lot before we add a validation command to add all current allocation users to Keycloak. I'll add the suggestion for now

knikolla replied:
Since we already have a lot of allocations and users of those, the validation command is still high priority otherwise only the new group memberships will be recorded and none of the previous ones.

knikolla commented on src/coldfront_plugin_cloud/kc_client.py:
```diff
@@ -0,0 +1,73 @@
+    def add_user_to_group(self, user_id, group_id):
+        url = f"{self.base_url}/admin/realms/{self.realm}/users/{user_id}/groups/{group_id}"
+        r = self.api_client.put(url)
+        r.raise_for_status()
```
If the user is already part of the group don't error out here, just log an info.

QuanMPhm replied:
From local testing, if the user is already part of the group, 204 is returned, so not error is raised.

knikolla commented on src/coldfront_plugin_cloud/kc_client.py:
```diff
@@ -0,0 +1,73 @@
+    def remove_user_from_group(self, user_id, group_id):
+        url = f"{self.base_url}/admin/realms/{self.realm}/users/{user_id}/groups/{group_id}"
+        r = self.api_client.delete(url)
+        r.raise_for_status()
```
If the user was not part of the group, don't error out here, just log an info.

QuanMPhm replied:
From local testing, it seems if the user is already not part of the group, `204` is returned, so not error is raised.

QuanMPhm commented:
@knikolla @naved001 I have responded to all comments

larsks commented on src/coldfront_plugin_cloud/kc_client.py:
```diff
@@ -0,0 +1,73 @@
+        session = requests.session()
+        session.headers.update(headers)
+        return session
+
```
@QuanMPhm This seems like a reasonable critique. Looking at the code in `src/coldfront_plugin_cloud/tasks.py`, it looks like there are many operations that will end up requesting a new api client multiple times. I would cache the result of `KeyCloakAPIClient.api_client`...

```python
    @functools.cached_property
    def get_api_client(self):
        params = {
            "grant_type": "client_credentials",
            "client_id": self.client_id,
            "client_secret": self.client_secret,
        }
        r = requests.post(self.token_url, data=params)
        r.raise_for_status()
        headers = {
            "Authorization": ("Bearer %s" % r.json()["access_token"]),
            "Content-Type": "application/json",
        }
        session = requests.session()
        session.headers.update(headers)
        return session
```

...and *remove* the cache from `tasks.get_kc_client`, so that (a) you're only calling `api_client` once per instance of `KeyCloakAPIClient`, and (b) every method in `tasks.py` will start with a fresh api client (because of the call to `get_kc_client`).

At that point, there's no need to `get_kc_client` so you can just write:

```python
    kc_admin_client = kc_client.KeyCloakAPIClient()
```

QuanMPhm replied:
I see. I'll look at it more closely, now that this PR is not due two days ago.

QuanMPhm replied:
@larsks Ok I see your idea now. Initially, my thought was, if we cache the class method, regardless of whether external functions that initialize the class is cached or not, the class function is "permanently" cached, across all instances of the class. Reading some stuff and testing lets me see your point

larsks commented on src/coldfront_plugin_cloud/kc_client.py:
```diff
@@ -0,0 +1,77 @@
+        """Return None if group not found"""
+        query = f"search={group_name}&exact=true"
+        url = f"{self.base_url}/admin/realms/{self.realm}/groups?{query}"
+        r = self.api_client.get(url)
```
Make python build urls for you:

```suggestion
        query = {
          "search": group_name,
          "exact": "true",
        }
        url = f"{self.base_url}/admin/realms/{self.realm}/groups"
        r = self.api_client.get(url, params=query)
```

larsks commented on ci/setup-keycloak.sh:
```diff
@@ -0,0 +1,61 @@
+
+set -xe
+
+sudo docker rm -f keycloak
```
@QuanMPhm why the thumbs down? This seems like a valid critique. A shell script with `set -e` active will exit if any command fails with an error.

QuanMPhm replied:
@larsks From local testing, running `docker rm` with the `-f` option makes the command return 0 if the container doesn't exist.

When I saw this comment, I remember we also have [a similar statement ](https://github.com/nerc-project/coldfront-plugin-cloud/blob/25a55a2bbcbf06163edb29e77c714516c48176fa/ci/microshift.sh#L14)in one of the other CI files, and quickly tested the command behavior.

larsks commented on src/coldfront_plugin_cloud/kc_client.py:
```diff
@@ -0,0 +1,77 @@
+        url = f"{self.base_url}/admin/realms/{self.realm}/groups?{query}"
+        r = self.api_client.get(url)
+        r.raise_for_status()
+        r_json = r.json()
```
Everywhere you are receiving a JSON response you should be using a model:

```
groups = GroupResponse.model_validate(r.json())
return groups[0].id
```

This will make it much more clear what data you're receiving from the API, and will make it easier to access, and it reduces the opportunities for typos.

larsks commented on ci/setup-keycloak.sh:
```diff
@@ -0,0 +1,61 @@
+    quay.io/keycloak/keycloak:25.0 start-dev
+
+# wait for keycloak to be ready
+until curl -s http://localhost:8080/auth/realms/master; do
```
This seems like a good idea.

QuanMPhm commented:
Addressed all comments, except [one](https://github.com/nerc-project/coldfront-plugin-cloud/pull/249#discussion_r3126172751) waiting for @knikolla response
